### PR TITLE
PP: All gather once at start

### PR DESCRIPTION
**Pull request by gobbleturk**
# Description

Add an option to all gather weights over FSDP initially before pipeline begins. This is a memory/time tradeoff - with this option on we perform less all gathers and reduce scatters but need to store the weights and gradients without FSDP sharding.

Prior to this change, PP + FSDP had the behavior of all gather and reduce scattering the weights every microbatch. After this change with `pipeline_fsdp_ag_once=True`, we will initially AG all the weights once and RS them once, which should be a factor of `num_pipeline_microbatches` less communication.

I found that the pipeline module had to be a submodule of a decoder so we could reference (and manipulate) self.layers.variables inside of the pipeline - hence some refactoring of how a pipeline module is initialized. Additionally it was tricky to construct the sharding annotations necessary to AG over FSDP - the weights don't exist yet until calling `.init` on pipeline module, so we can't pass the sharding annotations as a property until after calling .init. This PR instead plumbs in an optional sharding constraint to `__call__`.

# Tests
On my v4-8 devbox saw a 20% speedup with FSDP=2, PP=2 pdb=1, seq=2k. Will test this more at scale. Also added an additional test in pipeline_parallelism_test.py, although we were already confident this should pass as this PR only modifies a sharding constraint. Potentially we would replace this implementation with a shard mapped one in which case the unit test is even more useful

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed.

### CODE REVIEW

File: MaxText/layers/pipeline.py [hunk 1]
```diff
@@ -452,6 +570,7 @@ def __call__(
       positions: jnp.ndarray,
       deterministic: bool,
       model_mode=common_types.MODEL_MODE_TRAIN,
+      partition_spec=None,  # Pytree of sharding specifications of the weights (aka self.layers.variables)
```
Comment (RissyRan) [inline MaxText/layers/pipeline.py:573]: Oh nice. Does it works as expected?

**Review by RissyRan — COMMENTED** [review]
LGTM! Thanks Matt! At high level, have 2 questions:

1) have you already had a test to ensure the same logits with/without `pipeline_fsdp_ag_once`? Or I missed it somewhere?
2) Do you want to attach profiles of 2 implementation and show the AG differences?

File: MaxText/layers/pipeline.py (hunk 1 above)
Comment (gobbleturk) [reply MaxText/layers/pipeline.py:573]: So far yes! If we find it doesn't we can replace the sharding constraint implementation with a shard map

**Comment by gobbleturk** [comment]
> LGTM! Thanks Matt! At high level, have 2 questions:
> 
> 1. have you already had a test to ensure the same logits with/without `pipeline_fsdp_ag_once`? Or I missed it somewhere?
> 2. Do you want to attach profiles of 2 implementation and show the AG differences?

This PR includes a test in pipeline_parallelism_test which asserts that the output and gradients match a non-pipeline implementation.

I ran experiemnts on my v4-8 devbox with ici_pipeline_parallism=2 (fsdp=2) and pdb=2, seq=2k
* [trace](https://xprof.corp.google.com/trace_viewer/mattdavidow-5332985171866700224?hosts=t1v-n-5e50d777-w-0&host_index=0&trace_filter_config={}&view_start=-319.535&view_end=1422.812) with new feature false 
* [trace](https://xprof.corp.google.com/trace_viewer/mattdavidow-2261075635177290598?hosts=t1v-n-5e50d777-w-0&host_index=0&trace_filter_config={}&view_start=-171.229&view_end=1137.237) with  new feature true
